tests/check_pip

```diff
@@ -0,0 +1,197 @@
+		trap 'cleanup 2>/dev/null || rm -f ${LOCK_FILE} 2>/dev/null || true ; wait ; exit 143 ;' SIGTERM || EXIT_CODE=143
+		trap 'cleanup 2>/dev/null || rm -f ${LOCK_FILE} 2>/dev/null || true ; wait ; exit 131 ;' SIGQUIT || EXIT_CODE=131
+		# SC2173 - https://github.com/koalaman/shellcheck/wiki/SC2173
+		#trap 'cleanup 2>/dev/null || rm -f ${LOCK_FILE} 2>/dev/null || true ; wait ; exit 1 ;' SIGSTOP || EXIT_CODE=7
+		trap 'cleanup 2>/dev/null || rm -f ${LOCK_FILE} 2>/dev/null || true ; wait ; exit 130 ;' SIGINT || EXIT_CODE=130
+		trap 'cleanup 2>/dev/null || rm -f ${LOCK_FILE} 2>/dev/null || true || true ; wait ; exit 137 ;' SIGABRT || EXIT_CODE=137
+		trap 'cleanup 2>/dev/null || rm -f ${LOCK_FILE} 2>/dev/null || true ; wait ; exit ${EXIT_CODE} ;' EXIT || EXIT_CODE=1
+else
+		# shellcheck disable=SC2046
+		printf "\t%s\n" "Check Setup Scripts Tests Coverage already in progress by "$(head "${LOCK_FILE}") >&2 ;
+		exit 126 ;
+fi
+
+# check_pip: Audit and validate Python requirements files.
+
+# THIS IS THE ACTUAL TEST DIR USED (update _TEST_ROOT_DIR as needed)
+_TEST_ROOT_DIR=$(git rev-parse --show-toplevel 2>/dev/null) ;
+if [[ -d ../.git ]] ; then
+	_TEST_ROOT_DIR="../" ;
+elif [[ -d ./.git ]] ; then
+	_TEST_ROOT_DIR=$(pwd) ;
+elif [[ ( -d $(git rev-parse --show-toplevel 2>/dev/null) ) ]] ; then
+	_TEST_ROOT_DIR=$(git rev-parse --show-toplevel 2>/dev/null) ;
+else
+	printf "\t%s\n" "FAIL: missing valid repository or source structure" >&2 ;
+	EXIT_CODE=40
+fi
+
+cd $_TEST_ROOT_DIR 2>/dev/null || EXIT_CODE=40
+
+# Get a list of files to check using git ls-tree with filtering
+FILES_TO_CHECK=$(git ls-tree -r --full-tree --name-only HEAD -- **/requirements.txt *-requirements.txt requirements.txt 2>/dev/null || EXIT_CODE=3)
+
+# Enable auto-correction if '--fix' argument is provided
+if [[ "$1" == "--fix" ]]; then
+	AUDIT_OPTIONS="--fix --strict ${AUDIT_OPTIONS}"
+	printf "%s\n" "Auto-correction enabled."
+fi
+
+# THIS IS THE ACTUAL TEST
+# Iterate over files and run checks
+for req_file in $FILES_TO_CHECK; do
+	printf "%s\n" "Checking ${req_file}" ;
+	printf "%s\n" "Auditing ${req_file} for security vulnerabilities..."
+	{ pip-audit $AUDIT_OPTIONS "${req_file}" || EXIT_CODE=4 ;} ; wait ;
+	if [[ ("${EXIT_CODE}" -eq 0) ]] ; then
+		printf "%s\n" "Checking licenses in $req_file..." ;
+		{ pip install -r "$req_file" --quiet >/dev/null || EXIT_CODE=4 ;} 2>/dev/null ; wait ;
+		{ pip-licenses $LICENSE_OPTIONS || EXIT_CODE=5 ;} 2>/dev/null ; wait ;
+	else
+		printf "%s\n" "FAIL: Found requirements errors." >&2 ;
+	fi
+done
+
+# cleaning up and reporting
+
+if [[ ("${EXIT_CODE}" -eq 0) ]] ; then
+	printf "%s\n" "OK: Found no detected requirements errors." ;
+else
+	printf "%s\n" "FAIL: Found requirements errors." >&2 ;
+fi
+
+cleanup || rm -f ${LOCK_FILE} 2>/dev/null || : ;
+
+# unset when done
+unset _TEST_ROOT_DIR 2>/dev/null || : ;
+unset AUDIT_OPTIONS 2>/dev/null || : ;
+unset LICENSE_OPTIONS 2>/dev/null || : ;
+
+wait ;
+exit ${EXIT_CODE:-255} ;
```

Comment:
interesting, let's explore this further.

Reply:
🤩 wow, this honestly looks too good to believed. I vaguely recall bash functions having their own scope, will your approach actually work for things like signal handling, exiting (when in order to abort the whole script), ulimit and umask or will this actually just limit the impact to the function scope for those kinds of things? Can you show me an example to demonstrate the function and script scope concepts in bash?

Reply:
🤔 This answers my question but also leads to even more questions:
* regarding the refactoring of the functions `acquire_lock` and `handle_signals` your solution does not correctly solve the concurrency problem in the case where a SIGHUP (or similar signal) is received during the transition from returning from `acquire_lock`, and trapping from within `handle_signals` (consider if we add job throttling or suspend/pause signaling) this is why I used the if-else branching in the original implementation to ensure cleanup is called to cleanup from the lock. 
* regarding the refactoring of the functions `acquire_lock`, `handle_signals` and `cleanup` the order of definition seems odd; can bash really use functions defined later in the script? Where can I find the bash documentation supporting this?
* if `LOCK_FILE` is defined in `acquire_lock`
* did you overlook that the inconsistencies of `flock` availability is a major part of why a custom locking solution is used. See `CEP-4` in the comments of #124
* what happens scope-wise with the function's `export LC_CTYPE` will the script still restore the user's value of LC_CTYPE when it exits? This script should not change the parent/calling environment state of LC_CTYPE as this could be disastrous to users! Please demonstrate with examples (plural). Unless I'm wrong this should be changed to a script global variable. This bug was introduced by my original implementation. I'll add a fix to the batch of changes.

Reply:
🤔 where can I find official BASH documentation supporting your answer:
> In a shell script, exported variables affect the script and any child processes but do not modify the parent shell's environment after the script exits.

🙈 depending on that documentation we might actually be fine in this case.

Reply:
Ok, I've got an improved implementation in the works, I'll include it with the other batch of changes soon.